### ISSUE TITLE
C# support

### ISSUE THREAD

**Issue by mackowski** [semgrep/semgrep#1392]
I'm opening a ticket to indicate interest in support for the C# Language.
I can help with developing it if you can point me how to start and where to look.

**Comment by dlukeomalley** [comment]
Thank you for filing this @mackowski. Adding @aryx and @nbrahms to discuss what adding support for a new language looks like. We'd love to get more folks involved there.

**Comment by aryx** [comment]
Are you familiar with OCaml? the only remaining part to handle C# needs to be written in OCaml (in semgrep/semgrep-core/parsing/ like the Parse_ruby_tree_sitter.ml in this directory for example, but here it would be a Parse_csharp_tree_sitter.ml. We can generate the boilerplate for this file, but we still need to convert that in the generic AST (defined in pfff/h_program-lang/AST_generic.ml)

**Comment by mackowski** [comment]
I do not know OCaml, but it seems to be fun to learn it. Anyway, as I did not do it before I do not know how much time is needed to do it. I am ready to learn something new and help you with this project, but I also do not want to be a roadblock.
If you can generate the boilerplate in a new branch, I can try to start writing mapping, I will use other languages as examples. If you have capacity and willingness to do this faster do not wait for me, I will review the code to understand it better.

**Comment by aryx** [comment]
Sounds great! @mjambon can you add csharp in ocaml-tree-sitter-lang? I'll create the Parse_csharp_tree_sitter.ml boilerplate with a few cases filled in as a starting point.

**Comment by mjambon** [comment]
@aryx I just added csharp to ocaml-tree-sitter-lang. I parsed one sample program successfully on the first try but didn't run large-scale stats.

**Comment by aryx** [comment]
Ok,  I'll create the Parse_csharp_tree_sitter.ml boilerplate with a few cases filled in as a starting point.

**Comment by ievans** [comment]
@aryx looks like you added the basic parser; did you want to follow up with @mackowski ?

**Comment by aryx** [comment]
No I didn't. I guess I need to add a few simple cases to be a better starting point and maybe some documentation on where to find the relevant AST definitions.

**Comment by aryx** [comment]
The main work @mackowski  is to replace all the todo in https://github.com/returntocorp/semgrep/blob/develop/semgrep-core/parsing/Parse_csharp_tree_sitter.ml with the relevant construct defined in https://github.com/returntocorp/pfff/blob/master/h_program-lang/AST_generic.ml
Not trivial, but not impossible :)

**Comment by mackowski** [comment]
looks good! Next week I will be on a small trip without my laptop but after that I will start working on that :)

**Comment by Sjord** [comment]
I have [started an attempt](https://github.com/Sjord/semgrep/tree/csharp).

**Comment by Sjord** [comment]
I can't find anything in AST_Generic to represent an expression with an infix operator, such as `1 + 2`. I would expect `expr * operator * expr`. There is `AssignOp`, but as I understand, that is for `a += 1`. Am I missing something?

**Comment by aryx** [comment]
Yes you need to convert that in a Call (IdSpecial (Op Plus), e1, e2)
You can look in the codebase of pffff for js_to_generic.ml, or java_to_generic.ml to see how we convert those languages to the generic AST.

**Comment by aryx** [comment]
Feel free to ask many many questions @Sjord. We know AST_generic.ml is not very well commented, so we know it's tedious to write those converters from tree-sitter CST to AST_generic.ml
Fantastic job btw, this looks great.

**Comment by Sjord** [comment]
C# has [multidimensional arrays](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/multidimensional-arrays):

```csharp
        // Three-dimensional array.
        int[, ,] array3D = new int[,,] { { { 1, 2, 3 }, { 4, 5, 6 } },
                                        { { 7, 8, 9 }, { 10, 11, 12 } } };
        // The same array with dimensions specified.
        int[, ,] array3Da = new int[2, 2, 3] { { { 1, 2, 3 }, { 4, 5, 6 } },
                                            { { 7, 8, 9 }, { 10, 11, 12 } } };

        // Accessing array elements.
        System.Console.WriteLine(array2D[0, 0]);
```

I was thinking of representing this as a [TyArray](https://github.com/returntocorp/pfff/blob/master/h_program-lang/AST_generic.ml#L880) with a [Tuple](https://github.com/returntocorp/pfff/blob/master/h_program-lang/AST_generic.ml#L305), so that writing `array2D[0, 0]` is actually accessing `array2D` with the tuple (0, 0). 

However, this presents a problem when parsing something like this:

```csharp
int[, ,] array3D
```

This is an array with a tuple of (nothing, nothing, nothing), and that can't be represented as a Tuple as far as I know. Any ideas?

Edit: an alternative is to treat it as an array of arrays. C# also [has those](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/jagged-arrays), but maybe it's not so bad if we mix up jagged arrays with multidimensional arrays.

**Comment by aryx** [comment]
Hmm, yes not super easy to represent right now. We could extend the generic AST at some point to better handle those.
In the mean time when a construct does not fit, you can use one of the OtherXxx constructor and Ox_Todo,
for example here could return for those multidimensional array something like
OtherType (OT_Todo, [T (TyArray ((lbracket_tok, None, rbracket_tok), type_elt)])

And later when we improve the generic AST we can transform those OtherType in something more precise.

**Comment by aryx** [comment]
How are things progressing @Sjord ? Anything we can do to help? 
Do you think we could merge what you wrote?

**Comment by Sjord** [comment]
Progress is slow but steady. The parser can currently parse most common features. We started with 329 todo's. I have resolved 201 of those, and added another 30. The TODO's I added are mostly when I am not sure if what I'm doing is the correct way.

I ran into a couple of [upstream](https://github.com/tree-sitter/tree-sitter-c-sharp/issues/138) [issues](https://github.com/tree-sitter/tree-sitter-c-sharp/issues/141) in treesitter. The treesitter developers are very fast and helpful with resolving these, but I could use your help to get these fixes into semgrep-core. 

I am currently only working on the parser, and testing by running `semgrep-core -dump_ast`. Presumably other parts of semgrep also need some work before semgrep works on C#. I think semgrep currently doesn't even call this parser. I could use some help with that.

The parser is currently half-finished. We could merge it in the current state. It wouldn't break anything, but it is not really complete. Do you have a review process? Perhaps it is a good idea to start reviewing already, before the whole thing is done.

**Comment by aryx** [comment]
Fantastic! We can definitely help later to update to the latest tree-sitter-csharp and also to extend the grammar to allow semgrep-specific constructs (metavariables, ellipsis). In the mean time, I think the easiest way to go forward and for you to make a Pull Request on our repo so we can review it and merge it.

**Comment by aryx** [comment]
@dlukeomalley does Sjord need to sign some CLA agreement?

**Comment by Sjord** [comment]
In [`declaration_expression`, I return a PatVar](https://github.com/returntocorp/semgrep/blob/develop/semgrep-core/parsing/Parse_csharp_tree_sitter.ml#L1841-L1844), which needs a type:

```ocaml
PatVar of type_ * (ident * id_info) option
```

However, if you have C# like this:

```
foreach (var item in values)
```

The variable `item` doesn't really have a type. So far I have mapped `var` to None, where a type option is allowed.

Any ideas? Should I map `var item` to a PatId? Should I map `int item` to a PatVar, or a PatTyped PatId, or something else?

**Comment by aryx** [comment]
You can use PatId if the iterator variable does not have a type.

**Comment by Sjord** [comment]
> I ran into a couple of [upstream](https://github.com/tree-sitter/tree-sitter-c-sharp/issues/138) [issues](https://github.com/tree-sitter/tree-sitter-c-sharp/issues/141) in treesitter. The treesitter developers are very fast and helpful with resolving these, but I could use your help to get these fixes into semgrep-core. 

Could you help with this, @aryx or @mjambon ? I want to use the latest treesitter C# grammar in semgrep-core. Could you perhaps import that from upstream?

**Comment by aryx** [comment]
@mjambon can you update ocaml-tree-sitter-lang to use the latest tree-sitter-csharp?

**Comment by Sjord** [comment]
Most common language features are implemented in the treesitter parser. One feature that isn't implemented yet is LINQ queries. I intent to convert LINQ queries to function calls, to Select, OrderBy, etc. This is pretty hard. And I think I've run into an [upstream bug](https://github.com/tree-sitter/tree-sitter-c-sharp/issues/144) with this.

**Comment by aryx** [comment]
I still count around 60 todo().
Do you want us to finish those?

**Comment by Sjord** [comment]
No, I'll keep working on this. What I meant was that language features that are commonly used are implemented, and rarely used features are still to do. The result of this is that the parser in its current state can parse most files.

I am only working on [Parse_csharp_tree_sitter.ml](https://github.com/returntocorp/semgrep/blob/develop/semgrep-core/parsing/Parse_csharp_tree_sitter.ml), which does the the CST -> AST conversion. Once this parser is done, it would be nice if you can integrate it in semgrep, and finish the remainders for C# support in semgrep.

**Comment by aryx** [comment]
Ok sounds good.
@mjambon could you add the C# grammar in semgrep-grammars/ and start the extension to support $X (we can do ... later).
We also need to imitate the sgrep_spatch_pattern rule we have in the pfff grammars, because this is the first time we will rely on tree-sitter to also parse the pattern.

**Comment by aryx** [comment]
I've created a separate issue to use the tree-sitter parser to also parse the semgrep C# pattern.

**Comment by aryx** [comment]
FYI @Sjord, I made a big refactoring of AST_generic.stmt in https://github.com/returntocorp/semgrep/pull/2289 to make stmt a record. I made an helper function AST_generic.s to help build stmt from a stmt_kind.
I've refatored Parse_csharp_tree_sitter.ml but if you have some unpushed branch I suggest you to rebase from develop.

**Comment by Sjord** [comment]
How should async/await be mapped? I think async should be a keyword_attribute on methods, and await a special used in `Call`. Does that make sense?

**Comment by aryx** [comment]
there is a Await of tok * expr in AST_generic and an Async keyword_attribute

**Comment by aryx** [comment]
Ok we're getting really close to C# support in Semgrep!

**Comment by aryx** [comment]
Hopefully we can release in the next version of semgrep next week (0.37)

**Comment by aryx** [comment]
We are also at around 97% parsing success with tree-sitter, so not bad:
https://dashboard.semgrep.dev/metric/semgrep.core.c-sharp.parse.pct

**Comment by cfguimaraes** [comment]
@Sjord congratulations, you have done a lot, the community praises you.
I'm here to offer my help.
I see that the work you have done is far away from my knowledge, but I believe I can help with testing, or if some suggest, I could learn to help.

P.S. Sorry for my English, I'm brazilian guy that made a cheaper course :)

**Comment by Sjord** [comment]
Thanks, @cfguimaraes. You can help me improve the parser by creating small C# test files that have a construct that is currently not implemented. There are two ways to find those:

1. Search in [Parse_csharp_tree_sitter.ml](https://github.com/returntocorp/semgrep/blob/develop/semgrep-core/Parsing/tree_sitter/Parse_csharp_tree_sitter.ml) for `todo env`. The first occurence is [here](https://github.com/returntocorp/semgrep/blob/develop/semgrep-core/Parsing/tree_sitter/Parse_csharp_tree_sitter.ml#L259-L272), in a function called `attribute_target_specifier`. Search in [the docs how attribute target specifiers work](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/attributes), and create a little C# file similar [to the files here](https://github.com/returntocorp/semgrep/tree/develop/semgrep-core/tests/csharp/parsing).
2. Try to parse actual C# files until you encounter a file that can't be parsed correctly. Then reduce it to a minimal test case.

You can run `semgrep-core -dump_ast -lang cs somefile.cs` to check whether the file parses. What you are looking for is:

```
Fatal error: exception (Failure "not implemented")
```

If you have a minimal test case, this makes it easier for me to implement correct parsing.

What's also needed is semgrep integration tests, like [these](https://github.com/returntocorp/semgrep/tree/develop/semgrep-core/tests/python) but for C#. However, I think C# is currently not sufficiently implemented to make passable tests, and I am not really sure how to write these. @aryx, can you comment on that?

**Comment by aryx** [comment]
For the semgrep integration tests you can follow https://semgrep.dev/docs/status/#maturity-definitions and try at least the tests for the "Alpha" category. You can imitate the tests for lua/ which has currently the "Alpha" state.
If something does not work we probably need to extend the semgrep-grammar for C# at https://github.com/returntocorp/semgrep-grammars/blob/master/src/semgrep-c-sharp/grammar.js
(and again, to possibly imitate what we did for the other languages in his semgrep-grammars/src/semgrep-xxx directories.

**Comment by cfguimaraes** [comment]
I've been looking at these links, I believe I can help you right now with option 2, parsing files and making it reproducible.
As I learn more about Semgrep I can write tests, no problem =D
And who knows someday I could help you with the parser.

Talk to you as soon as I found this exception in our codebase (is plenty of scenarios)

**Comment by ievans** [comment]
@cfguimaraes I am excited that you are interested in helping! If you would like even faster responses, you can also join us on Slack: https://r2c.dev/slack. Most of the developers are there and we are very friendly!

**Comment by Sjord** [comment]
We really need the latest tree-sitter-c-sharp to make semgrep work correctly, even for alpha functionality. I started updating it in [semgrep-grammars](https://github.com/returntocorp/semgrep-grammars), but this broke the build after the PR was merged, and I don't fully understand why.

**Comment by aryx** [comment]
What is the error message? What is broken?

**Comment by Sjord** [comment]
The CircleCI image [here](https://github.com/returntocorp/semgrep-grammars) showed failed instead of passed, but when I tried to reproduce it I got different errors each time. However, I think I solved it [in my PR](https://github.com/returntocorp/semgrep-grammars/pull/20).

**Comment by Sjord** [comment]
@mjambon, could you perhaps import the latest [upstream](https://github.com/tree-sitter/tree-sitter-c-sharp) C# grammar? The version in [semgrep-c-sharp](https://github.com/returntocorp/semgrep-c-sharp) is pretty old and contains several serious bugs.

**Comment by aryx** [comment]
@mjambon I can have a go at it. Good way to test for me the new ocaml-tree-sitter organization.

**Comment by cfguimaraes** [comment]
Hello all :)

@Sjord when running locally in windows via docker I have got these exceptions:


```
FormaPagamentoService.cs:1:0: Fatal Error: Common.Impossible

Fatal error: exception Common.Impossible
error invoking semgrep with:
        /usr/local/bin/semgrep-core -lang cs -dump_ast FormaPagamentoService.cs
        Command '['/usr/local/bin/semgrep-core', '-lang', 'cs', '-dump_ast', 'FormaPagamentoService.cs']' returned non-zero exit status 2.
An error occurred while invoking the semgrep engine; please help us fix this by creating an issue at https://github.com/returntocorp/semgrep
PagamentoCartaoService.cs:1:0: Fatal Error: AST_generic_helpers.NotAnExpr

Fatal error: exception AST_generic_helpers.NotAnExpr
error invoking semgrep with:
        /usr/local/bin/semgrep-core -lang cs -dump_ast PagamentoCartaoService.cs
        Command '['/usr/local/bin/semgrep-core', '-lang', 'cs', '-dump_ast', 'PagamentoCartaoService.cs']' returned non-zero exit status 2.
An error occurred while invoking the semgrep engine; please help us fix this by creating an issue at https://github.com/returntocorp/semgrep
```

I know this isn't what you are want me to look for `Fatal error: exception (Failure "not implemented")`. But I'm curious if I should ignore or report these as separate?

**Comment by Sjord** [comment]
> Fatal error: exception Common.Impossible

This is probably because of [a bug](https://github.com/tree-sitter/tree-sitter-c-sharp/issues/141) in the tree-sitter grammar. It doesn't currently make much sense to create a test case for this, until we have an updated grammar.

> Fatal error: exception AST_generic_helpers.NotAnExpr

This is new to me. This could be interesting to have a test case for.

**Comment by cfguimaraes** [comment]
I found what is the cause of the Error you are interested in.

I've made the minimum reproducible code that I can

```csharp
Example
{
    public class ExampleClass
    {
        public async Task<ApiResponseTransacao> Method(Request request)
        {
            (int x, int y) = request.A.B().Value;
        }
    }
}
```

> Sorry for the ugly names, it's because of my NDA.

In the code above Request is a DTO Class, request.A is a string property and request.B is an extension method that returns an `(int,int)?` Nullable tuple.

I want to know from you what is the best way you would like to receive these snippets. And of course, 

My next step is learn more about @aryx suggestion on https://semgrep.dev/docs/status/#maturity-definitions to known a little more about definitions and after ~copy some implementation of integration tests~ write integration tests :)

@Sjord if you want, I can make a project with this code.

**Comment by cfguimaraes** [comment]
Important too, I've been running docker this way
`docker run --rm -v "${PWD}:/src" returntocorp/semgrep --dump-ast -l cs file.cs`

**Comment by Sjord** [comment]
Thanks, @cfguimaraes. Yes, I think a new GitHub project with code snippets is the easiest way. Then I will add them to the semgrep parsing tests as soon as I implement the parsing code for them.

**Comment by aryx** [comment]
update to latest tree-sitter-csharp here: https://github.com/returntocorp/ocaml-tree-sitter/pull/154

**Comment by aryx** [comment]
And here is the PR to update semgrep: https://github.com/returntocorp/semgrep/pull/2567 (incomplete).

**Comment by aryx** [comment]
@Sjord there are lots of changes in the latest tree-sitter-csharp. Would you mind fix #2567 ?

**Comment by cfguimaraes** [comment]
@Sjord sorry for the late reply.
I see that you already have evolved since our last talk.
I've submitted the samples to GitHub but didn't mentioned here.

The link with samples is https://github.com/cfguimaraes/semgrep-AST_generic_helpers.NotAnExpr

**Comment by Sjord** [comment]
#3121

**Comment by mschwager** [comment]
Can we close this out now that C# has reached alpha: https://github.com/returntocorp/semgrep/pull/3121? If we'd like to track additional improvements we should create new, specific issues (e.g. beta, GA, parse errors, etc).

**Comment by mschwager** [comment]
Closing out - feel free to open new, specific issues for additional C# work!

Thanks for all the hard work here!
